Use machines instead of placements

jat-canonical opened the pull request:
## Done

* Add support for placing applications on same machines 
* Add support for deploying the terraform plan in the CI

## JIRA / Launchpad bug

Fixes [#AC-3318](https://warthogs.atlassian.net/browse/AC-3318)

jat-canonical committed:
chore: bump juju provider version requirement to 0.19

jat-canonical committed:
feat: use juju machines for optimal placement of applications

This commit adds support for spinning up machines and placing multiple
applications on the same machine as our bundles do currently.

jat-canonical committed:
feat(subcluster): use a dedicated ca for etcd

As etcd is authenticated through mTLS, we need a separate CA for issuing
certs to etcd for enhanced security.

jat-canonical committed:
test(subcluster): add assertions for juju machines for subcluster

jat-canonical committed:
test(controller): add assertions for verifying the use of juju machines

jat-canonical committed:
feat(gh): add test to deploy terraform plan in CI

This commit adds a job in the workflow to test the full deployment for
the plan in CI with the LXD provider.

ajanon reviewed (approved):
LGTM

morphis commented on .github/workflows/build-and-test.yaml:
```diff
@@ -31,9 +31,13 @@ jobs:
-    runs-on: [self-hosted, linux, ARM64, jammy, medium]
+    runs-on: [self-hosted, linux, X64, jammy, large]
+    env:
+      UBUNTU_PRO_TOKEN: ${{ secrets.UBUNTU_PRO_TOKEN }}
```
As mentioned I think we can not use the pro token by requiring 1.26 which drops the explicit requirement to have pro enabled on the machine when the charm is deployed

jat-canonical committed:
feat!: remove pro token requirement from the plan

The charms no longer have a hard requirement for the presence of a pro
token necessarily. The plan should deploy the charms without a token and
still go through.

jat-canonical committed:
fix(gh): use edge track for anbox channel

jat-canonical committed:
fix(gh): fail build workflow on test failures

jat-canonical committed:
chore(registry): update juju terraform provider version to 0.19.0

morphis commented on variables.tf:
```diff
@@ -1,15 +1,15 @@
+//     condition     = length(var.ubuntu_pro_token) > 0
+//     error_message = "Ubuntu Pro Token should not be empty"
+//   }
+// }
```
Is this expected? Why is this commented?

jat-canonical replied:
I was testing out dropping it completely. I'll remove this after the workflow passes without it.

morphis commented:
```
model "anbox-subcluster-a" found, waiting...
ERROR timed out waiting for "anbox-subcluster-a" to reach goal state
ERROR timed out waiting for "anbox-controller" to reach goal state
All models deployed successfully
```

@jat-canonical looks like something is still wrong here.

jat-canonical commented:
> ```
> model "anbox-subcluster-a" found, waiting...
> ERROR timed out waiting for "anbox-subcluster-a" to reach goal state
> ERROR timed out waiting for "anbox-controller" to reach goal state
> All models deployed successfully
> ```
> 
> @jat-canonical looks like something is still wrong here.

i know that is why I am fixing it. I'll move the PR to draft for the time being

jat-canonical committed:
fix(controller): change the endpoint used by the new CA charm

The self-signed-certificate uses a different endpoint instead of the old
easyrsa charm.

jat-canonical committed:
fix(controller): use edge channel for nats charm

The latest changes to use `self-signed-certificates` charm as the CA
charm broke the deployment as NATS charm only support the given charm on
edge channel as of now.

morphis reviewed (approved):
LGTM
